pkg/operator/ceph/file/mds.go

```diff
@@ -14,194 +14,189 @@ See the License for the specific language governing permissions and
-	_, err = context.Clientset.ExtensionsV1beta1().Deployments(fs.Namespace).Create(deployment)
-	if err != nil {
-		if !errors.IsAlreadyExists(err) {
-			return fmt.Errorf("failed to create mds deployment. %+v", err)
-		}
-		logger.Infof("mds deployment %s already exists", deployment.Name)
-	} else {
-		logger.Infof("mds deployment %s started", deployment.Name)
+// return true if an attempt was made to prepare the filesystem associated for upgrade
+func (c *cluster) deleteLegacyMdsDeployment() bool {
+	legacyName := fmt.Sprintf("%s-%s", AppName, c.fs.Name) // rook-ceph-mds-<fsname>
+	logger.Debugf("getting legacy mds deployment %s for filesystem %s if it exists", legacyName, c.fs.Name)
+	_, err := c.context.Clientset.Extensions().Deployments(c.fs.Namespace).Get(legacyName, metav1.GetOptions{})
+	if err != nil && errors.IsNotFound(err) {
+		logger.Infof("legacy mds deployment %s not found, no update needed", legacyName)
+		return false
+		// if any other error, failed to get legacy dep., but it exists, so still try deleting it.
 	}
-
-	return nil
+	logger.Infof("deleting legacy mds deployment %s", legacyName)
+	if err := mdsdaemon.PrepareForDaemonUpgrade(c.context, c.fs.Namespace, c.fs.Name, fsUpgradePrepareTimeout); err != nil {
+		logger.Errorf("failed to prepare filesystem %s for mds upgrade while deleting legacy deployment %s, continuing anyway: %+v", c.fs.Name, legacyName, err)
+	}
+	if err := k8sutil.DeleteDeployment(c.context.Clientset, c.fs.Namespace, legacyName); err != nil {
+		logger.Errorf("failed to delete legacy deployment %s, USER should delete the legacy mds deployment manually if it still exists: %+v", legacyName, err)
+	}
+	return true
 }
 
-// Delete the file system
-func DeleteFilesystem(context *clusterd.Context, fs cephv1beta1.Filesystem) error {
-	// Delete the mds deployment
-	k8sutil.DeleteDeployment(context.Clientset, fs.Namespace, instanceName(fs))
+func (c *cluster) start() error {
+	// If attempt was made to prepare daemons for upgrade, make sure that an attempt is made to
+	// bring fs state back to desired when this method returns with any error or success.
+	var fsPreparedForUpgrade = false
+	defer func() {
+		if fsPreparedForUpgrade {
+			if err := mdsdaemon.FinishedWithDaemonUpgrade(c.context, c.fs.Namespace, c.fs.Name, c.fs.Spec.MetadataServer.ActiveCount); err != nil {
+				logger.Errorf("for filesystem %s, USER should make sure the Ceph fs max_mds property is set to %d: %+v",
+					c.fs.Name, c.fs.Spec.MetadataServer.ActiveCount, err)
+			}
+		}
+	}()
+	// Delete legacy daemonset, and remove it if it exists.
+	// This is relevant to the upgrade from Rook 0.8 to 0.9.
+	fsPreparedForUpgrade = c.deleteLegacyMdsDeployment()
+
+	// Always create double the number of metadata servers to have standby mdses available
+	replicas := c.fs.Spec.MetadataServer.ActiveCount * 2
+	for i := 0; i < int(replicas); i++ {
```

Comment:
If the admin increases or decreases the `activeCount` in the filesystem CRD, does the operator add/remove the individual deployments? I see it will likely add more, but not remove during a decrease.